[Example] Llama 4 maverick with checkpointing (#6273)
skypilot-org/skypilot: llm/llama-4-finetuning/README.md

Meta's Llama 4 represents the next generation of open-source large language models, featuring advanced capabilities with the **Llama-4-Maverick-17B-128E model** - a 400B  parameter (17B active) Mixture of Experts (MoE) architecture with 128 experts.

> This guide shows how to use [SkyPilot](https://github.com/skypilot-org/skypilot) and various frameworks to **finetune Llama 4 on your own infra**. Everything is packaged in simple [SkyPilot YAMLs](https://docs.skypilot.co/en/latest/getting-started/quickstart.html), that can be launched with one command on your infra: 

Reviewer: ```suggestion
This guide shows how to use [SkyPilot](https://github.com/skypilot-org/skypilot) with torchtune and Llama Factory to **finetune Llama 4 on your own infra**. Everything is packaged in simple [SkyPilot YAMLs](https://docs.skypilot.co/en/latest/getting-started/quickstart.html), that can be launched with one command on your infra: 
```

Author: Updated

- Kubernetes cluster
- Cloud accounts ([16+ clouds supported](https://docs.skypilot.co/en/latest/getting-started/installation.html))

> ## 📁 Available Configuration Files

Reviewer: ```suggestion
## 📁 Available Recipes
```

> Choose the right configuration for your needs:

Reviewer: ```suggestion
Choose the right recipe for your needs:
```

> | **Configuration File** | **Requirements** | **Description** |

Reviewer: ```suggestion
| **Recipe** | **Requirements** | **Description** |
```

|------------------------|------------------|-----------------|

> | 🌟 **llama-4-maverick-sft.yaml** | **4 nodes**<br>16x H200 GPUs<br>1000+ GB CPU memory | **✅ RECOMMENDED** - CPU-friendly full finetuning using LLaMA-Factory with CPU offloading. Best starting point for most users. |

Reviewer: This doesn't use `LLaMA-Factory` anymore, right? Also maybe rephrase to `Full finetuning using torchtune with CPU offloading. Best starting point for most users.`

> | 🌟 **llama-4-maverick-sft.yaml** | **4 nodes**<br>16x H200 GPUs<br>1000+ GB CPU memory | **✅ RECOMMENDED** - CPU-friendly full finetuning using LLaMA-Factory with CPU offloading. Best starting point for most users. |

Reviewer: ```suggestion
| 🌟 **llama-4-maverick-sft.yaml** | **4 nodes**<br>32x H200 GPUs<br>1000+ GB CPU memory per node | **✅ RECOMMENDED** - CPU-friendly full finetuning using LLaMA-Factory with CPU offloading. Best starting point for most users. |
```

Author: Updated

| 🎯 **llama-4-maverick_gke_llama_factory_lora.yaml** | **2 nodes**<br>16x H100 GPUs<br>1000+ GB CPU memory | **Memory efficient** - LoRA fine-tuning with lower resource requirements. Great for limited GPU resources. |

> ## 🚀 Recommended: CPU-Friendly Full Finetuning

Reviewer: ```suggestion
## Full finetuning with CPU offloading 
```

> **Start here for most users!** This approach uses [LLaMA-Factory](https://github.com/hiyouga/LLaMA-Factory) with CPU offloading to reduce GPU memory requirements.

Reviewer: Change to torchtune?

> **SkyPilot YAML**: [`llama-4-maverick-sft.yaml.yaml`](llama-4-maverick-sft.yaml.yaml)

Reviewer: ```suggestion
**SkyPilot YAML**: [`llama-4-maverick-sft.yaml`](llama-4-maverick-sft.yaml)
```

Author: Nice catch.

  --rdzv_endpoint=$MASTER_ADDR:29500 \
  full_finetune_distributed \
  --config llama4/maverick_17B_128E_full \
  model_dir=/tmp/Llama-4-Maverick-17B-128E-Instruct \
  dataset.packed=True tokenizer.max_seq_len=4096 \
  gradient_accumulation_steps=1 \
  enable_activation_offloading=True \
  activation_offloading_use_streams=False \
  optimizer_in_bwd=True \
  optimizer=torch.optim.AdamW \
  optimizer_kwargs.fused=True \
  max_steps_per_epoch=1 \
  epochs=10 \
  enable_dcp=True \
  enable_async_checkpointing=True \
  resume_from_checkpoint=False \
  keep_last_n_checkpoints=1 \
  fsdp_cpu_offload=True
```
    
</details>

```bash
sky launch -c maverick-torchtune llama-4-maverick-sft.yaml \
  --env HF_TOKEN
```

## Alternative Approaches

### LoRA Fine-tuning (Lower Resource Requirements)
For users with limited GPU resources, LoRA (Low-Rank Adaptation) provides an efficient alternative:

```bash
# LoRA finetuning - requires fewer resources

> sky launch -c maverick-lora llama-4-maverick-llama-factory-lora.yaml \

Reviewer: YAML path probably needs to be updated?

Author: Updated all references to point to correct yaml
